packages/gatsby-plugin-utils/src/test-plugin-options-schema.ts

```diff
@@ -0,0 +1,35 @@
+import Joi from "joi"
+import { ObjectSchema } from "./utils/plugin-options-schema-joi-type"
+
+type PluginSchemaFunction = ({ Joi }) => ObjectSchema
+
+export function testPluginOptionsSchema<PluginOptions = object>(
+  pluginSchemaFunction: PluginSchemaFunction,
+  pluginOptions: PluginOptions
+): Array<string> {
+  const pluginOptionsNames = Object.keys(pluginOptions)
+  const pluginSchema = pluginSchemaFunction({ Joi })
+  const errors: Array<string> = []
+
+  pluginOptionsNames.forEach(pluginOptionName => {
+    const partialSchema = pluginSchema.extract(pluginOptionName)
+
+    const { error } = partialSchema.validate(pluginOptions[pluginOptionName], {
+      abortEarly: false,
+    })
+
+    if (error) {
+      const errorMessage = error.message
+
+      // In the case of an array, "value" does not exist in the error message
+      // and so we can't replace it with the plugin option name, we have to concat it
+      const message = errorMessage.includes(`"value"`)
+        ? errorMessage.replace(`"value"`, `"${pluginOptionName}"`)
+        : `"${pluginOptionName}" ${errorMessage}`
+
+      errors.push(message)
+    }
+  })
+
+  return errors
```

Comment:
Maybe we should return something better here in case the options are valid or rename the function to something including "errors" or return a boolean or something? 🤔 

It's not very obvious based on this test that the options are valid:

```JS
it('should not error with valid options', () => {
  expect(testPluginOptionsSchema(schema, { trackingId: "bla" })).toEqual([]);
})
```

Reply:
I agree with that. I have two suggestion:

- the one you're suggesting: including "error" in the name that would imply that we only want to verify for errors, not for the validity of a schema

- Accepting to return an object from the function. This way folks are able to choose what makes sense for them

`const { isValid, errors } = testPluginOptionsSchema(schema, { trackingId: "bla" })`

Reply:
I like the second option a lot! That'll make it very clear when reading the tests what the person is testing 👍 

```JS
const { isValid, errors } = testPluginOptionsSchema(schema, { trackingId: "bla" })

expect(isValid).toEqual(true)
expect(errors).toEqual([])

const { isValid, errors } = testPluginOptionsSchema(schema, { trackingId: undefined })

expect(isValid).toEqual(false)
expect(errors).toMatchInlineSnapshot()
```